libs/s25main/Loader.cpp

```diff
@@ -930,140 +949,156 @@ bool Loader::MergeArchives(libsiedler2::Archiv& targetArchiv, libsiedler2::Archi
+        throw LoadError(libsiedler2::getErrorString(ec));
     }
+
     logger_.write(_("done in %ums\n")) % duration_cast<milliseconds>(timer.getElapsed()).count();
 
-    return true;
+    return archive;
 }
 
 /**
- *  @brief Lädt eine Datei in ein Archiv.
- *
- *  @param[in] pfad    Der Dateipfad
- *  @param[in] palette (falls benötigt) die Palette.
- *  @param[in] archiv  Das Zielarchivinfo.
+ *  @brief Load a single file into the archive
  *
- *  @return @p true bei Erfolg, @p false bei Fehler.
+ *  @param[in] archiv Archive to add file to
+ *  @param[in] filePath Path to file
+ *  @param[in] palette palette to use if required
  */
-bool Loader::LoadArchiv(libsiedler2::Archiv& archiv, const std::string& pfad, const libsiedler2::ArchivItem_Palette* palette)
+libsiedler2::Archiv Loader::DoLoadFile(const boost::filesystem::path& filePath, const libsiedler2::ArchivItem_Palette* palette)
 {
     const Timer timer(true);
 
-    std::string file = config_.ExpandPath(pfad);
-
-    logger_.write(_("Loading \"%s\": ")) % file;
+    logger_.write(_("Loading \"%s\": ")) % filePath;
     fflush(stdout);
 
-    if(int ec = libsiedler2::Load(file, archiv, palette))
+    libsiedler2::Archiv archive;
+    if(int ec = libsiedler2::Load(filePath.string(), archive, palette))
     {
         logger_.write(_("failed: %1%\n")) % libsiedler2::getErrorString(ec);
-        return false;
+        throw LoadError(libsiedler2::getErrorString(ec));
     }
 
     logger_.write(_("done in %ums\n")) % duration_cast<milliseconds>(timer.getElapsed()).count();
 
-    return true;
+    return archive;
 }
 
-bool Loader::LoadOverrideDirectory(const std::string& path)
+bool Loader::LoadOverrideDirectory(const bfs::path& path)
 {
     if(!bfs::is_directory(path))
     {
         logger_.write(_("Directory does not exist: %s\n")) % path;
         return false;
     }
 
-    // yes, load all files in the directory
     const Timer timer(true);
 
-    logger_.write(_("Loading LST,LBM,BOB,IDX,BMP,TXT,GER,ENG,INI files from \"%s\"\n")) % config_.ExpandPath(path);
+    logger_.write(_("Loading LST,LBM,BOB,IDX,BMP,TXT,GER,ENG,INI files from \"%s\"\n")) % path;
 
-    std::vector<std::string> filesAndFolders = ListDir(path, "lst", true);
-    filesAndFolders = ListDir(path, "lbm", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "bob", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "idx", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "bmp", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "txt", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "ger", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "eng", true, &filesAndFolders);
-    filesAndFolders = ListDir(path, "ini", true, &filesAndFolders);
+    std::vector<bfs::path> filesAndFolders;
+    for(const auto ext : {"lst", "lbm", "bob", "idx", "bmp", "txt", "ger", "eng", "ini"})
+    {
+        std::vector<bfs::path> curFiles = ListDir(path, ext, true);
```

Comment:
result can be const (or simply const auto?)

Reply:
Good catch!

I'd only use auto where the type is mostly clear from the RHS or complicated (iterators) as IMO the type is a vital part for knowing the code.

Reply:
yes I often do that too, but at least the const ;-)

Reply:
Done: https://github.com/Return-To-The-Roots/s25client/pull/1242/commits/c7e4b87cf09eb1d50f0fe7a2e50f518cae0690eb